rust-demangle.c

```diff
@@ -1147,27 +1152,170 @@ static void demangle_const_str_literal(struct rust_demangler *rdm) {
+                PRINT("::");
+                ident.ascii += 2;
+                ident.ascii_len -= 2;
+            } else {
+                PRINT(".");
+                ident.ascii += 1;
+                ident.ascii_len -= 1;
+            }
+        } else if (ident.ascii[0] == '$') {
+            const char *end_ptr =
+                (const char *)memchr(&ident.ascii[1], '$', ident.ascii_len - 1);
+            if (!end_ptr)
+                break;
+            const char *escape = &ident.ascii[1];
+            size_t escape_len = end_ptr - escape;
+
+            if (escape_len == 2 && escape[0] == 'S' && escape[1] == 'P') {
+                PRINT("@");
+            } else if (escape_len == 2 && escape[0] == 'B' && escape[1] == 'P') {
+                PRINT("*");
+            } else if (escape_len == 2 && escape[0] == 'R' && escape[1] == 'F') {
+                PRINT("&");
+            } else if (escape_len == 2 && escape[0] == 'L' && escape[1] == 'T') {
+                PRINT("<");
+            } else if (escape_len == 2 && escape[0] == 'G' && escape[1] == 'T') {
+                PRINT(">");
+            } else if (escape_len == 2 && escape[0] == 'L' && escape[1] == 'P') {
+                PRINT("(");
+            } else if (escape_len == 2 && escape[0] == 'R' && escape[1] == 'P') {
+                PRINT(")");
+            } else if (escape_len == 1 && escape[0] == 'C') {
+                PRINT(",");
+            } else {
+                if (escape[0] != 'u') {
+                    break;
+                }
+
+                const char *digits = &escape[1];
+                size_t digits_len = escape_len - 1;
+
+                bool invalid = false;
+                for (size_t i = 1; i < digits_len; i++) {
+                    if (!IS_DIGIT(digits[i]) &&
+                        !(digits[i] >= 'a' && digits[i] <= 'f')) {
+                        invalid = true;
+                        break;
+                    }
+                }
+                if (invalid)
+                    break;
+
+                struct hex_nibbles hex;
+
+                hex.nibbles = digits;
+                hex.nibbles_len = digits_len;
+
+                uint32_t c = 0;
+                for (size_t i = 0; i < hex.nibbles_len; i++)
+                    c = (c << 4) | decode_hex_nibble(hex.nibbles[i]);
+
+                if (!(c < 0xd800 || (c > 0xdfff && c < 0x10ffff))) {
+                    break; // Not a valid unicode scalar
+                }
+
+                if (c >= 0x20 && c <= 0x7e) {
+                    // Printable ASCII
+                    char v = (char)c;
+                    print_str(rdm, &v, 1);
+                } else {
+                    // FIXME show printable unicode characters without hex
+                    // encoding
+                    PRINT("\\u{");
+                    char s[9] = {0};
+                    sprintf(s, "%" PRIx32, c);
+                    PRINT(s);
+                    PRINT("}");
+                }
```

Comment:
Hmm, this could probably just reuse the `char` printing helper?

Reply:
The `print_quoted_escaped_char` function will put quotes around the char.

Reply:
I was referring to whatever is used for each Unicode character of a constant string literal.

Reply:
`demangle_const_str_literal` uses `print_quoted_escaped_char` except with `"` instead of `'` as argument for the "quote" parameter. Neither option is entirely correct for identifiers.